examples/vision/gradient_centralization.py

```diff
@@ -0,0 +1,253 @@
+
+
+"""
+Rescale and augment the data
+"""
+
+train_ds = prepare(train_ds, shuffle=True, augment=True)
+test_ds = prepare(test_ds)
+
+"""
+# Define a model
+
+In this section we will define a Convolutional neural network
+"""
+
+model = tf.keras.Sequential(
+    [
+        layers.Conv2D(16, (3, 3), activation="relu", input_shape=(300, 300, 3)),
+        layers.MaxPooling2D(2, 2),
+        layers.Conv2D(32, (3, 3), activation="relu"),
+        layers.Dropout(0.5),
+        layers.MaxPooling2D(2, 2),
+        layers.Conv2D(64, (3, 3), activation="relu"),
+        layers.Dropout(0.5),
+        layers.MaxPooling2D(2, 2),
+        layers.Conv2D(64, (3, 3), activation="relu"),
+        layers.MaxPooling2D(2, 2),
+        layers.Conv2D(64, (3, 3), activation="relu"),
+        layers.MaxPooling2D(2, 2),
+        layers.Flatten(),
+        layers.Dropout(0.5),
+        layers.Dense(512, activation="relu"),
+        layers.Dense(1, activation="sigmoid"),
+    ]
+)
+
+"""
+We will also create a callback which allows us to easily measure the total training time
+and the time taken for each epoch  since we are interested in comparing the effect of
+Gradient Centralization on the model we built above.
+"""
+
+
+class TimeHistory(tf.keras.callbacks.Callback):
+    def on_train_begin(self, logs={}):
+        self.times = []
+
+    def on_epoch_begin(self, batch, logs={}):
+        self.epoch_time_start = time()
+
+    def on_epoch_end(self, batch, logs={}):
+        self.times.append(time() - self.epoch_time_start)
+
+
+"""
+# Train the model without GC
+"""
+
+"""
+We now train the model we built earlier without Gradient Centralization which we can
+compare to the training performance of the model trained with GGradient Centralization.
+"""
+
+time_callback_no_gc = TimeHistory()
+model.compile(
+    loss="binary_crossentropy",
+    optimizer=RMSprop(learning_rate=1e-4),
+    metrics=["accuracy"],
+)
+
+history_no_gc = model.fit(
+    train_ds, epochs=10, verbose=1, callbacks=[time_callback_no_gc]
+)
+
+"""
+# Train the model with GC 
+
+We will now train the same model, this time using Gradient Centralization. We use the
+package
+[`gradient-centralization-tf`](https://github.com/Rishit-dagli/Gradient-Centralization-Ten
+sorFlow/) package to help us implement Gradient Centralization on top of our optimizer.
+We will make use of the `gctf.centralized_gradients_for_optimizer()` method to do so.
+This method accepts a `tf.keras.optimizers.Optimizer` object and modifies our optimizer
+to use Gradient Centralization.
+
+The experiments in [this paper](https://arxiv.org/abs/2004.01461) on various
+applications, including general image classification, fine-grained image classification,
+detection and segmentation and Person ReID demonstrate that GC can consistently improve
+the performance of DNN learning.
+"""
+
+optimizer = RMSprop(learning_rate=1e-4)
+optimizer.get_gradients = gctf.centralized_gradients_for_optimizer(optimizer)
```

Comment:
I see the implementation is pretty short: https://github.com/Rishit-dagli/Gradient-Centralization-TensorFlow/blob/a7c5226dad86ca42341061e3fafc8c8d1ec3f51f/gctf/centralized_gradients.py#L5-L42

I recommend packaging the implementation directly in the example. The goal of a code example is to explain how to implement something, not just demonstrate how to use an external package.

The easiest way to implement this would be to create a subclass of `RMSprop` that overrides the `get_gradients` method. You can make it very short by leveraging `super().get_gradients()`.

Reply:
Sure thing, that makes a lot of sense, thanks!
I added that but did not use `super().get_gradients()` since I did not want to run the `get_gradients()` function first.